lib/actions/write.js

```diff
@@ -1,104 +1,96 @@
+    const fields = cfg.order.split(`${delimiter}`).map(x => { return x.trim() })
+    const orderedData = data.map((value) => {
+      const result = fields.map((key) => {
+        const filtered = value[key]
+        return filtered
+      })
+      return result
+    })
+    csvString = papa.unparse({
+      fields,
+      data: orderedData
+    }, unparseOptions)
+  } else {
+    csvString = papa.unparse(data, unparseOptions)
   }
 
-  if (timeout) {
-    clearTimeout(timeout);
+  if (!cfg.uploadToAttachment) {
+    await this.emit('data', messages.newMessageWithBody({ csvString }))
+    this.logger.info(`Complete, memory used: ${process.memoryUsage().heapUsed / 1024 / 1024} Mb`)
+    return
   }
 
-  timeout = setTimeout(async () => {
-    readyFlag = false;
-
-    self.logger.info('Closing the stream due to inactivity');
-
-    const finalRowCount = rowCount;
-    self.logger.info('The resulting CSV file contains %s rows', finalRowCount);
-    ax.put(putUrl, stringifier, {
-      method: 'PUT',
-      timeout: REQUEST_TIMEOUT,
-      retry: REQUEST_MAX_RETRY,
-      delay: REQUEST_RETRY_DELAY,
-      maxContentLength: REQUEST_MAX_CONTENT_LENGTH,
-    });
-    stringifier.end();
+  const attachmentProcessor = new AttachmentProcessor()
+  let attachment
+  try {
+    attachment = await attachmentProcessor.uploadAttachment(csvString)
+  } catch (err) {
+    this.logger.error(`Upload attachment failed: ${err}`)
+    this.emit('error', `Upload attachment failed: ${err}`)
+  }
+  const body = {
+    attachmentUrl: attachment.config.url,
+    type: '.csv',
+    size: Buffer.byteLength(csvString),
+    attachmentCreationTime: new Date(),
+    attachmentExpiryTime: new Date(),
+    contentType: 'text/csv'
+  }
+  const respData = messages.newMessageWithBody(body)
+
+  respData.attachments = {
+    'data.csv': {
+      'content-type': body.contentType,
+      size: body.size,
+      url: body.attachmentUrl
+    }
+  }
 
-    const messageToEmit = messages.newMessageWithBody({
-      rowCount: finalRowCount,
-      url: signedUrl.get_url,
-    });
-    const fileName = `${messageToEmit.id}.csv`;
-    messageToEmit.attachments[fileName] = {
-      'content-type': 'text/csv',
-      url: signedUrl.get_url,
-    };
-    signedUrl = null;
-    rowCount = 0;
-    self.logger.trace('Emitting message');
-    await self.emit('data', messageToEmit);
-    await init(cfg);
-  }, TIMEOUT_BETWEEN_EVENTS);
+  await this.emit('data', respData)
+  this.logger.info(`Complete, memory used: ${process.memoryUsage().heapUsed / 1024 / 1024} Mb`)
+  this.logger.info('Attachment created successfully')
+}
 
-  let row = msg.body.writer;
-  if (cfg.writer.columns) {
-    const columns = Object.keys(_.keyBy(cfg.writer.columns, 'property'));
-    row = _.pick(row, columns);
+async function writeCSV(msg, cfg) {
+  const { body } = msg
+  cfg.header = body.header
+
+  // if not array - create array from all fn calls and send data to proceedData
+  if (Array.isArray(body.items)) {
```

Comment:
I'm not a big fan of using the structure of the input to determine behavior, but what you have seems like it would work.

Reply:
Would this be better option?

Create 2 different files:
- writeStream.js
- writeArray.js

call them from **component.json** for separate action, each file will call **write.js**, but with predefined option isArray (can also add checks)

Reply:
I think that approach would be slightly cleaner.

Reply:
Done